### PR TITLE
Generalization of SampleLogsBehaviour (MergeRuns, ConjoinXRuns)

### LINKED ISSUE #23947: Parameter names in conflict when using MergeRuns and ConjoinXRuns
Parameters defined in the IPF for using `MergeRuns` cause conflicts when stitching using the same set of parameters with `ConjoinXRuns`:

A real world example run gives:
```
Stitch1DMany-[Notice] Stitch1DMany started
Stitch1D-[Notice] Scale Factor Y is: 0.0228918 Scale Factor E is: 0.00044508
ConjoinXRuns-[Error] Could not join workspace: __Stitch1D_intermediate_workspace_2__. Reason: "Item "CollAngle.actual_coll_angle" has different values in workspaces! Found: 3.7620000839233398 in workspace __Stitch1D_intermediate_workspace_2__ but the value in the first workspace was: 0.61299997568130493.
ConjoinXRuns-[Error] ". Skipping.
Stitch1DMany-[Notice] Stitch1DMany successful, Duration 0.05 seconds
```

Parameter names of the IPF are hard-coded in `SampleLogsBehaviour`.
Several solutions exist to overcome the problem.

(It may be good to print units in error messages in general)

### PR DESCRIPTION
**Pull request by vreimund**
**Description of work.**

Currently, parameters of the parameter file specified by inflexible naming convention in `SampleLogsBehaviour.cpp` are used for `MergeRuns` and `ConjoinXRuns`. This may cause conflicts, if merging and joining is desired. This can be the case when running a reduction work-flow. For instance, `MergeRuns` parameters were specified and `Stitch1D` (uses `ConjoinXRuns` for point data) skipped the second to stitch workspace since the collimation angle was not identical in case of FIGARO data as shown in the issue. Thus, the behavior of merging runs and stitching or joining work-spaces does not agree.
This PR generalizes the algorithm `SampleLogsBehaviour` and adds new tests.

Default parameter names (formerly only parameter names) are used for `MergeRuns`
Prefix `conjoin_` + default parameter names are used for `ConjoinXRuns`
New algorithm can make use of new parameter names.

The user must modify the parameters file for changing and controlling the stitch behaviour since override properties are not exposed.

Please note code changes with regard to using structs containing strings and using shared MatrixWorkspace pointers in public methods. 

**To test:**

Tests are in place. Tests of `MergeRuns`, `ConjoinXRuns` did not change from a functional point of view.
Make sure the new `SampleLogsBehaviourTest` tests the new behaviour.

Fixes #23947.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by reimundILL** [comment]
I added in progress since I fix the missing units in the sample logs for `MergeRuns` etc and modified the tests. Via tests, I found problems to be solved.

File: Framework/Algorithms/src/ConjoinXRuns.cpp [hunk 1]
```diff
@@ -333,20 +333,26 @@ void ConjoinXRuns::exec() {
...
+  logEntries.sampleLogsList = getPropertyValue(SampleLogsBehaviour::LIST_PROP);
+  logEntries.sampleLogsWarn = getPropertyValue(SampleLogsBehaviour::WARN_PROP);
+  logEntries.sampleLogsWarnTolerances =
+      getPropertyValue(SampleLogsBehaviour::WARN_TOL_PROP);
+  logEntries.sampleLogsFail = getPropertyValue(SampleLogsBehaviour::FAIL_PROP);
+  logEntries.sampleLogsFailTolerances =
+      getPropertyValue(SampleLogsBehaviour::FAIL_TOL_PROP);
   const std::string sampleLogsFailBehaviour = getProperty("FailBehaviour");
+  SampleLogsBehaviour::ParameterName parName = {};
+  parName.SUM_MERGE = "conjoin_sample_logs_sum";
```
Comment (DTasev) [inline Framework/Algorithms/src/ConjoinXRuns.cpp]: Same for `conjoin log strings` - please move them to const strings and reuse in test, reduces the change of errors due to spelling mistakes

**Review by dtasev — CHANGES_REQUESTED** [review]
Heading in a better direction by storing the log names in a struct - a few comments added

File: Framework/Algorithms/inc/MantidAlgorithms/RunCombinationHelpers/SampleLogsBehaviour.h [hunk 2]
```diff
@@ -51,22 +56,33 @@ class MANTID_ALGORITHMS_DLL SampleLogsBehaviour {
...
+    std::string sampleLogsWarnTolerances;
+    std::string sampleLogsFail;
+    std::string sampleLogsFailTolerances;
+  };
+
+  SampleLogsBehaviour(API::MatrixWorkspace_sptr ws, Kernel::Logger &logger,
+                      const SampleLogNames &logEntries = {"", "", "", "", "",
+                                                          "", ""},
+                      const ParameterName &parName = {
+                          "sample_logs_sum", "sample_logs_time_series",
```
[the remark this answers is not in the document]
Comment (reimundILL) [reply Framework/Algorithms/inc/MantidAlgorithms/RunCombinationHelpers/SampleLogsBehaviour.h]: I now tried to address your suggestions.

File: Framework/Algorithms/test/MergeRunsTest.h [hunk 3]
```diff
@@ -1102,46 +1102,46 @@ class MergeRunsTest : public CxxTest::TestSuite {
...
+    const std::string mergeType = MergeRuns::FAIL_MERGE;
     auto ws = create_group_workspace_with_sample_logs<double>(
         mergeType, "prop1", 1.0, 2.0, 0.0, 0.0);
     do_test_mergeSampleLogs(ws, "prop1", mergeType, "1", 1);
   }
 
   void
   test_mergeSampleLogs_fail_where_params_are_different_but_inside_tolerance_succeeds() {
-    std::string mergeType = SampleLogsBehaviour::FAIL_MERGE;
+    std::string mergeType = MergeRuns::FAIL_MERGE;
```
Comment (DTasev) [inline Framework/Algorithms/test/MergeRunsTest.h]: Can this `std::string` and below also be made `const`?

**Review by dtasev — CHANGES_REQUESTED** [review]
Minor comment about making more `std::string`s const in `MergeRunsTest`. 

Windows build seems to fail with unresolved external symbols errors.

Apart from this I think this PR is looking better, and nearly ready

File: Framework/Algorithms/test/MergeRunsTest.h (hunk 3 above)
Comment (reimundILL) [reply Framework/Algorithms/test/MergeRunsTest.h]: Good idea.

**Comment by DTasev** [comment]
Windows still doesn't like the symbols :(

**Comment by reimundILL** [comment]
Windows does not seem to see the cpp definition of the static variables. I plan to create a new namespace MergeRunsOptions
```
namespace MergeRunsOptions{
/// MergeRuns parameter names of the paramter file for sample log merging
static const std::string SUM_MERGE = "sample_logs_sum";
static const std::string TIME_SERIES_MERGE = "sample_logs_time_series";
static const std::string LIST_MERGE = "sample_logs_list";
static const std::string WARN_MERGE = "sample_logs_warn";
static const std::string WARN_MERGE_TOLERANCES =
    "sample_logs_warn_tolerances";
static const std::string FAIL_MERGE = "sample_logs_fail";
static const std::string FAIL_MERGE_TOLERANCES =
    "sample_logs_fail_tolerances";
}
```

**Comment by reimundILL** [comment]
MergeRuns uses DLLExport whereas SampleLogsBehaviour uses MANTID_ALGORITHMS_DLL.
Can this make the difference?

**Comment by DTasev** [comment]
In `DllConfig.h` this macro definition is active, which makes me think they're the same:

```cpp
#ifdef IN_MANTID_ALGORITHMS
#define MANTID_ALGORITHMS_DLL DLLExport
#else
...
```

Your latest changes seem to have fixed the build on Windows, the failure on Ubuntu Python 3 is unrelated I think

**Review by dtasev — APPROVED** [review]

**Pull request merged**